### PR TITLE
Implement binding redefinition

### LINKED ISSUE #198: Binding redefinition mechanism for REPL context
Blocks #197, #112
[TODO(#198)](https://github.com/morganey-lang/Morganey/search?utf8=%E2%9C%93&q=%22TODO%28%23198%29%22&type=)

### PR DESCRIPTION
**Pull request by rexim**
Close #198

### CODE REVIEW

File: src/main/scala/me/rexim/morganey/Main.scala [hunk 1]
```diff
@@ -78,6 +78,7 @@ object Main extends SignalHandler {
           case Success(ReplResult(newContext, message)) =>
             globalContext = newContext
             message.foreach(con.println)
+            con.println(newContext.bindings.map(_.variable.name).toString)
```
Comment (rexim) [inline src/main/scala/me/rexim/morganey/Main.scala]: Remove that! It's a troubleshooting leftover.

File: src/test/scala/me/rexim/morganey/interpreter/ReplContextSpecs.scala [hunk 2]
```diff
@@ -22,6 +22,18 @@ class ReplContextSpecs extends FlatSpec with Matchers with TestTerms with Mockit
...
   }
 
+  it should "keep the last unique binding" in {
+    def binding(v1: LambdaVar, v2: LambdaVar): MorganeyBinding =
+      MorganeyBinding(v1, I(v2))
+
+    val context = ReplContext(List(binding(z, z)))
+
+    context.addBinding(binding(x, x)).bindings should be (List(binding(x, x), binding(z, z)))
+    context.addBinding(binding(x, x)).addBinding(binding(x, y)).bindings should be (List(binding(x, y), binding(z, z)))
```
Comment (rexim) [inline src/test/scala/me/rexim/morganey/interpreter/ReplContextSpecs.scala]: Word wrap this line too

File: src/main/scala/me/rexim/morganey/interpreter/ReplContext.scala [hunk 3]
```diff
@@ -11,12 +11,11 @@ object ReplContext {
 
 case class ReplContext(bindings: List[MorganeyBinding] = Nil) {
   def addBinding(binding: MorganeyBinding): ReplContext = {
-    // TODO(#198): Binding redefinition mechanism for REPL context
-    ReplContext(binding :: bindings)
+    ReplContext(binding :: bindings.filter(_.variable != binding.variable))
```
Comment (rexim) [inline src/main/scala/me/rexim/morganey/interpreter/ReplContext.scala]: It's `O(n)`. I think it can be way faster. Create an issue for that.

**Review by ForNeVeR — APPROVED** [review]

**Review by keddelzz — APPROVED** [review]

File: src/main/scala/me/rexim/morganey/interpreter/ReplContext.scala [hunk 4]
```diff
@@ -10,13 +10,15 @@ object ReplContext {
 }
 
 case class ReplContext(bindings: List[MorganeyBinding] = Nil) {
+  // TODO(#360): Improve time asymptotic of add binding to REPL context
```
Comment (rexim) [inline src/main/scala/me/rexim/morganey/interpreter/ReplContext.scala]: the add binding operation*

File: src/main/scala/me/rexim/morganey/interpreter/ReplContext.scala [hunk 5]
```diff
@@ -10,13 +10,15 @@ object ReplContext {
...
+  //
+  // Right now it's O(N), but it can be imporved
```
Comment (rexim) [inline src/main/scala/me/rexim/morganey/interpreter/ReplContext.scala]: improved*

**Comment by rexim** [comment]
@ForNeVeR @keddelzz thank you guys for the review!

**Pull request merged**